### PR TITLE
remove this weird special case from promotion

### LINKED ISSUE #67534: Promotion creates invalid constants that are never used in invalid ways
Namely, we can have promoteds with interior mutable types but in an immutable allocation.
#67000 makes this explicit by introducing the `ignore_interior_mut_in_const_validation` field on the const interner.

We should instead rewrite such promoteds to just contain the parts of the promoted that is accessed. So

```rust
let x: &'static i32 = &[(Cell::new(42), 1), Cell::new(42), 2)][runtime_index].1;
```

currently produces

```rust
const PROMOTED: &[(Cell<i32>, i32); 2] = &[(Cell::new(42), 1), Cell::new(42), 2)];
let x: &'static i32 = &PROMOTED[runtime_index].1;
```

And we should be producing

```rust
const PROMOTED: &[i32; 2] = &[1, 2];
let x: &'static i32 = &PROMOTED[runtime_index];
```

This is definitely a nontrivial transformation, especially since it needs to be done on the generic promoted and not the computed final value (generic e.g. if we had `T::CELL_VALUE` instead of `Cell::new(42)`).
The reason it needs to be done on the MIR is that we also need to change the projection for obtaining the value of `x`, which we have no control over if we did it on the final constant.

### PR DESCRIPTION
**Pull request by RalfJung**
Promotion has a special case to ignore interior mutability under some specific circumstances. The purpose of this PR is to figure out what changes if we remove that. Since `Cell::new` and friends only get promoted inside `const`/`static` initializers these days, it actually is not easy to exploit this case: you need something like
```rust
const TEST_INTERIOR_MUT: () = {
    // The "0." case is already ruled out by not permitting any interior mutability in `const`.
    let _val: &'static _ = &(Cell::new(1), 2).1;
};
```

I assume something like `&Some(&(Cell::new(1), 2).1)` would hit the nested case inside `validate_rvalue`... though I am not sure why that would not just trigger nested promotion, first promoting the inner reference and then the outer one?

Fixes https://github.com/rust-lang/rust/issues/67534 (by simply rejecting that code^^)

r? @oli-obk (but for now this is not meant to be merged!)
Cc @rust-lang/wg-const-eval

### CODE REVIEW

File: src/test/ui/consts/promote-not.rs [hunk 1]
```diff
@@ -32,4 +34,14 @@ const TEST_UNION: () = {
...
+// In a `const`, we do not promote things with interior mutability. Not even if we "project it away".
+const TEST_INTERIOR_MUT: () = {
+    // The "0." case is already ruled out by not permitting any interior mutability in `const`.
+    let _val: &'static _ = &(Cell::new(1), 2).1; //~ ERROR temporary value dropped while borrowed
+};
+
+fn main() {
+    // We must not promote things with interior mutability. Not even if we "project it away".
+    let _val: &'static _ = &(Cell::new(1), 2).0; //~ ERROR temporary value dropped while borrowed
+    let _val: &'static _ = &(Cell::new(1), 2).1; //~ ERROR temporary value dropped while borrowed
```
Comment (RalfJung) [inline src/test/ui/consts/promote-not.rs:46]: These two already failed before.

**Review by oli-obk — COMMENTED** [review]
Is this maybe a leftover of the fact that promotion checking and const checking used to be one huge spaghetti monster? Or is it "just because" our promotion rules in const contexts kind of follow explicit promotion rules?

**Comment by RalfJung** [comment]
Ah, here is another case that is affected:
```rust
const FIVE: Cell<i32> = Cell::new(5);

#[inline(never)]
fn tuple_field() -> &'static u32 {
    // This test is MIR-borrowck-only because the old borrowck
    // doesn't agree that borrows of "frozen" (i.e., without any
    // interior mutability) fields of non-frozen temporaries,
    // should be promoted, while MIR promotion does promote them.
    &(FIVE, 42).1
}
```
(I wish we would have sticked to what the old borrowck allows.^^)

File: compiler/rustc_mir/src/transform/promote_consts.rs [hunk 2]
```diff
@@ -692,28 +667,7 @@ impl<'tcx> Validator<'_, 'tcx> {
...
-                        if ty.is_freeze(self.tcx.at(DUMMY_SP), self.param_env) {
-                            has_mut_interior = false;
-                            break;
-                        }
-
-                        place_projection = proj_base;
-                    }
-                }
+                let has_mut_interior = self.qualif_local::<qualifs::HasMutInterior>(place.local);
                 if has_mut_interior {
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply compiler/rustc_mir/src/transform/promote_consts.rs:671]: Yeah this is not meant to be clean code, just the easiest patch that works and passes CI and rustfmt.^^

File: compiler/rustc_mir/src/transform/promote_consts.rs [hunk 3]
```diff
@@ -326,41 +326,16 @@ impl<'tcx> Validator<'_, 'tcx> {
...
+                            return Err(Unpromotable);
                         }
 
                         // FIXME(eddyb) this duplicates part of `validate_rvalue`.
+                        let has_mut_interior =
+                            self.qualif_local::<qualifs::HasMutInterior>(place.local);
                         if has_mut_interior {
                             return Err(Unpromotable);
                         }
-                        if self.qualif_local::<qualifs::NeedsDrop>(place.local) {
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply compiler/rustc_mir/src/transform/promote_consts.rs:361]: To make the block of code that is the same as `validate_rvalue` one contiguous block, starting at the "FIXME this is duplicated".

**Comment by RalfJung** [comment]
> Is this maybe a leftover of the fact that promotion checking and const checking used to be one huge spaghetti monster? Or is it "just because" our promotion rules in const contexts kind of follow explicit promotion rules?

I think it is a "because we can" that was added when promotion was ported to MIR (probably it was natural to add given that it was the same code as const checking, where such precision is probably useful and less of a problem).

**Comment by oli-obk** [comment]
a check-only crater run should suffice, please feel free to start it whenever you think the PR is ready. I really hope we can get away with this change

**Comment by RalfJung** [comment]
eh...
@bors try

**Comment by RalfJung** [comment]
@craterbot check

**Comment by RalfJung** [comment]
All of these regressions are false positives. :)

**Comment by RalfJung** [comment]
I rebased and made CI pass. This is ready for FCP.

**Comment by joshtriplett** [comment]
:tada:
@rfcbot merge

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by RalfJung** [comment]
Rebased on newer master.

@oli-obk this PR is now awaiting your review. :)

**Pull request merged**